### PR TITLE
Yahoo amp-ad plugin

### PR DESCRIPTION
**Pull request by src-code**
Initial support for Yahoo display and Gemini ads via amp-ad plugin.

### CODE REVIEW

**Comment by src-code** [comment]
@djamrozik Daniel, any chance you can sign the Google CLA so we can move forward here?

File: ads/yahoo.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+*/
+
+import {validateData, loadScript} from '../3p/3p';
+
+function done (global, data) {
+    window.ampAdExecute(global, data);
+}
+
+export function yahoo (global, data) {
+    var cb = done.bind(this, global, data);
```
Comment (lannka) [inline ads/yahoo.js]: we use 2-space indentation

File: ads/_config.js [hunk 2]
```diff
@@ -335,6 +335,14 @@ export const adConfig = {
 
   widespace: {},
 
+  yahoo: {
+    prefetch: [
+      'https://s.yimg.com/os/ampad/display.js',
+      'https://s.yimg.com/os/ampad/gemini.js'
```
Comment (lannka) [inline ads/_config.js]: please append trailing `,`

File: ads/_config.js [hunk 3]
```diff
@@ -335,6 +335,14 @@ export const adConfig = {
...
+    ],
+    preconnect: 'https://us.adserver.yahoo.com'
```
Comment (lannka) [inline ads/_config.js]: same here

File: examples/ads.amp.html [hunk 4]
```diff
@@ -841,6 +842,14 @@ <h2 id="widespace">Widespace 300x50 Ad</h2>
     <div fallback></div>
   </amp-ad>
 
+  <h2 id="yahoo">Yahoo</h2>
+  <amp-ad width="300" height="250"
+      type="yahoo"
+      data-sid="954014446"
+      data-site="news"
+      data-sa='{"LREC":"300x250","secure":"true","content":"no_expandable;"}'>
+  </amp-ad>
```
Comment (lannka) [inline examples/ads.amp.html]: Could you also add an example for `adtype=gemini`

File: ads/yahoo.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+* @param {!Window} global
+* @param {!Object} data
+*/
```
Comment (lannka) [inline ads/yahoo.js]: pls move the JS doc to the right place

**Comment by src-code** [comment]
@lannka Since our ad implementation will live externally from this code, I believe we'll have to adopt the `renderStart`/`noContentAvailable` APIs in those scripts.  That said, 
It's unclear to me that we can make use of those APIs for our initial display ad implementation, since we'll be nesting an iframe inside amp-ad's iframe, and won't necessary have access to the relevant state.

By the way, the documentation for `renderStart` references `waitForRenderStart` list in `_config.js` which no longer exists - I assume that whitelist is no longer necessary?

File: ads/yahoo.js (hunk 5 above)
Comment (src-code) [inline ads/yahoo.js]: Went ahead and restructured this code a bit.

Comment (src-code) [inline ads/yahoo.js]: Done

File: examples/ads.amp.html (hunk 4 above)
Comment (src-code) [inline examples/ads.amp.html]: Done

File: ads/_config.js (hunk 3 above)
Comment (src-code) [inline ads/_config.js]: Done

File: ads/_config.js (hunk 2 above)
Comment (src-code) [inline ads/_config.js]: Done

File: ads/yahoo.js (hunk 1 above)
Comment (src-code) [inline ads/yahoo.js]: Updated

**Comment by src-code** [comment]
Hi @lannka - Checking in, anything further need to be done for this one?

File: ads/yahoo.js [hunk 6]
```diff
@@ -20,12 +20,12 @@ import {validateData, loadScript} from '../3p/3p';
 * @param {!Window} global
 * @param {!Object} data
 */
-export function yahoo (global, data) {
-  var cb = (function (global, data) {
+export function yahoo(global, data) {
+  const cb = (function(global, data) {
```
Comment (jridgewell) [inline ads/yahoo.js]: This bound function isn't necessary:

```js
function cb() {
  window.ampAdExecute(global, data);
}

// or
const cb = () => {
  window.ampAdExecute(global, data);
};

File: examples/ads.amp.html [hunk 7]
```diff
@@ -870,6 +871,21 @@ <h2 id="widespace">Widespace 300x50 Ad</h2>
...
 
+  <h2 id="yahoo">Yahoo Display</h2>
+  <amp-ad width="300" height="250"
+      type="yahoo"
+      data-sid="954014446"
+      data-site="news"
+      data-sa='{"LREC":"300x250","secure":"true","content":"no_expandable;"}'>
+  </amp-ad>
+
+  <h2 id="yahoo-gemini">Yahoo Gemini</h2>
```
Comment (lannka) [inline examples/ads.amp.html]: id is not needed here. Each ad network will have one ID for navigation.

File: examples/ads.amp.html [hunk 8]
```diff
@@ -870,6 +871,21 @@ <h2 id="widespace">Widespace 300x50 Ad</h2>
     <div fallback></div>
   </amp-ad>
 
+  <h2 id="yahoo">Yahoo Display</h2>
+  <amp-ad width="300" height="250"
+      type="yahoo"
+      data-sid="954014446"
+      data-site="news"
+      data-sa='{"LREC":"300x250","secure":"true","content":"no_expandable;"}'>
```
Comment (lannka) [inline examples/ads.amp.html]: please also add the following tags inside `amp-ad`
```
    <div placeholder></div>
    <div fallback></div>
```

**Comment by lannka** [comment]
@src-code render-start is not available in nested iframe yet. Will let you know once its ready.
The `waitForRenderStart` list is replaced by the `renderStartImplemented: true` in `_config.js`. Documentation was updated.

Thanks for your patience. Please address the comments and I will take another look.

**Comment by src-code** [comment]
@lannka Thanks, I'll update with fixed examples.

Though for the Gemini case, I likely will just pull that out.  Gemini is a little different in that it's a native ad which we're rendering inside content populated with amp-list.  amp-ad feels like the wrong approach for these ads, since we've already got the ad rendering available, so there's no reason to inject an iframe there (and we want to share styling with the content.)  Really the only reason I can find to use amp-ad is for the ad viewability tracking, which doesn't seem to be solvable with the recently-added viewability triggers supported by amp-analytics.  I'd hate to use amp-ad as just a tracking pixel in this case, or sticking the whole ad into the placeholder/fallback nodes and letting amp-ad fail silently (if that'd even work?)  Curious to know if you've seen other use cases such as this, and have a recommendation for how to best handle this sort of native advertising?

**Comment by lannka** [comment]
Using `amp-list` to display ads might be a mis-use of the widget. @jasti @cramforce, is A4A our final answer to these kind of requirements?

**Comment by src-code** [comment]
@lannka In our case, the ads are co-mingled in the feed that returns article data, which is why they end up rendered by amp-list.  In a sense they're just sponsored articles, but with the additional requirement of viewability tracking.  There's no need for an ad call since we already have all the data available to us, and we want the rendering to mostly match the other non-ad content.  I'm not sure A4A helps for those reasons?

**Comment by src-code** [comment]
In a sense, I think what I want is the ability to pass in the rendered ad via a child element such as what's done for `placeholder` or `fallback`.  The only problem then would be viewability tracking, since I wouldn't have access to the `observeIntersection` APIs to help me in that case.  And the `amp-analytics` viewability tracking is very limited (only `amp-` elements, selectors must be tag names or ids, and only fire once per selector, etc), so doesn't seem to be helpful here.

**Comment by jasti** [comment]
@src-code @lannka  why not submit an embed similar to what Taboola does?
http://output.jsbin.com/lenubepone Is it possible to get just the sponsored articles part of a separate call? This way, you'll get observerIntesection API too.
If you just want to use amp-list, that's fine but we'll consider it to be just part of the content and there will no special treatment.

**Comment by src-code** [comment]
@jasti We could certainly go the embed route - I'll explore that next, and if it seems promising, open a separate PR.  Meanwhile, I'm going to clean this one up a bit.

**Comment by src-code** [comment]
@laanka @jasti Actually, now that I think of it, going the embed route would not really help us much - we want to track the visibility of individual stories, not the aggregate.  The sponsored stories are intermixed with the non-sponsored stories, so we want to track each sponsored story independently...  So it seems like the only way to do that is to make each sponsored story its own amp-ad/amp-embed, but then we're back to the problem of having to feed pre-rendered content to the amp-ad via properties, or using amp-ad only as a tracking pixel - both of which seem wrong.

It'd be great if amp-ad would support a rendering option that didn't require an iframe, so we could just wrap existing, simple ad markup that doesn't require its own javascript, only than the amp-ad plugin code to track and beacon visibility.  Does that make sense, and would it seem possible?  Or is there a better way?

**Comment by jasti** [comment]
@src-code Not convinced using amp-ad/embed for each sponsored story is a bad idea. I don't think we could render a non A4A ad without an iframe.

**Comment by src-code** [comment]
@jasti Does A4A work outside of iframes?  I was under the impression that it must be a fully-formed AMP document, which seems to imply iframes?  There's no reason these ads couldn't be A4A, since they're just plain markup without javascript, but if they have to be iframed then it doesn't seem to help me much. My primary goal is to avoid iframes so I can take advantage of existing page styling, but I still need viewability tracking.

Maybe what's needed then is an enhancement to amp-analytics viewability tracking?  Having the ability to track these elements in a more fine-grained fashion (eg by class rather than id or tag, and firing a separate beacon for each), even for content injected late via amp-list, would be sufficient for what I'm trying to do...

**Comment by lannka** [comment]
@src-code A4A is in a same-origin iframe. 

If you put `amp-ad` as an item of `amp-list`, you should be able to track its visibility using `amp-analytics`, no matter A4A or not.

For CSS sharing, @dvoytenko do we currently have a way to pass CSS to A4A?

**Comment by src-code** [comment]
@lannka 

> If you put amp-ad as an item of amp-list, you should be able to track its visibility using amp-analytics, no matter A4A or not.

As I understand it, amp-analytics will only match on id or tag name, so there's no good way to match a set of ads of a particular type, as I'd need here (I'd want to write a selector like `amp-ad.gemini`).  Also, it'll only fire 1 beacon for all matching elements, rather than 1 beacon per element, I believe?  And I can't select by id since I don't know what ids will be present on the page until after the amp-list renders.

But it's a moot issue, because if I'm going to use an amp-ad plugin anyway, the viewability tracking becomes easy since I can use the `observeIntersection` API.  But then, styling becomes more difficult, since I'm stuck in an iframe.

**Comment by dvoytenko** [comment]
@lannka 

> For CSS sharing, @dvoytenko do we currently have a way to pass CSS to A4A?

Sure. The HTML is allowed to contain `<style amp-custom>...</style>` just like a normal AMP document.

**Comment by src-code** [comment]
In some ways, it seems like PR #5541 is similar to what I'd want (no iframe on amp-ad) but I'd want to avoid the network request (I already have the markup) and I'd want to track the visibility somehow.

**Comment by lannka** [comment]
@src-code #5541 will only work on pure image ad. 

I see you want to do two things:
- CSS sharing:  I would go for A4A 
- an enhanced amp-analytics selector: ask @avimehta :-)

**Comment by src-code** [comment]
@lannka My ads are almost pure image ads - they just have a couple lines of text in addition to a single image.  So the need is quite similar, I think.  

Actually I have another proposal for a change to amp-ad I am working on, and will open an issue to discuss.  I hacked around in amp-ad yesterday and got it working pretty easily, and seems to solve my issues without having go overcomplicate things with A4A and amp-analytics.  Will post more soon...

**Comment by src-code** [comment]
This PR diverged a bit with all the discussion about native ads, so I'm going to close it out in favor of a fresh PR dedicated solely to Yahoo display ads.

**Pull request closed without merging**